backends/bmv2/synthesizeValidField.cpp

```diff
@@ -0,0 +1,151 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+#include "synthesizeValidField.h"
+
+#include <boost/range/combine.hpp>
+#include <boost/range/irange.hpp>
+
+#include "backends/bmv2/helpers.h"
+#include "frontends/common/resolveReferences/referenceMap.h"
+#include "frontends/p4/methodInstance.h"
+#include "frontends/p4/typeChecking/typeChecker.h"
+#include "frontends/p4/typeMap.h"
+#include "ir/ir.h"
+
+namespace BMV2 {
+
+/// Adds a valid bit to all header types.
+class AddValidField final : public Modifier {
+ public:
+    AddValidField() { setName("AddValidField"); }
+ private:
+    bool preorder(IR::Type_Header* header) override {
+        const cstring validField = V1ModelProperties::validField;
+        if (header->getField(validField) != nullptr) return true;
+        auto field = new IR::StructField(validField, IR::Type::Bits::get(1));
+        header->fields.push_back(field);
+        return true;
+    }
+};
+
+/// Replaces calls to isValid() with references to the header's valid bit.
+class RewriteIsValidCalls final : public Transform {
+ public:
+    RewriteIsValidCalls(P4::ReferenceMap* refMap, P4::TypeMap* typeMap)
+            : refMap(refMap), typeMap(typeMap) {
+        CHECK_NULL(refMap); CHECK_NULL(typeMap);
+        setName("RewriteIsValidCalls");
+    }
+
+ private:
+    const IR::Node* postorder(IR::MethodCallExpression* expr) override;
+
+    P4::ReferenceMap* refMap;
+    P4::TypeMap* typeMap;
+};
+
+const IR::Node*
+RewriteIsValidCalls::postorder(IR::MethodCallExpression* expr) {
+    auto instance = P4::MethodInstance::resolve(expr, refMap, typeMap);
+    if (!instance->is<P4::BuiltInMethod>()) return expr;
+    auto builtin = instance->to<P4::BuiltInMethod>();
+    if (builtin->name != IR::Type_Header::isValid) return expr;
+
+    // The isValid() builtin is defined for both headers and header unions, but
+    // only headers have a valid bit - header unions work differently
+    // internally - so we'll leave isValid() invocations on header unions as-is.
+    auto type = typeMap->getType(builtin->appliedTo, true);
+    if (type->is<IR::Type_HeaderUnion>()) return expr;
+    BUG_CHECK(type->is<IR::Type_Header>(),
+              "Invoking isValid() on unexpected type %1%", type);
+
+    // If isValid() is being used as a table key element, it already behaves
+    // like a bit<1> value; just replace it with a reference to the valid bit.
+    const cstring validField = V1ModelProperties::validField;
+    auto member = new IR::Member(expr->srcInfo, builtin->appliedTo, validField);
+    if (getParent<IR::KeyElement>() != nullptr) return member;
+
+    // In other contexts, rewrite isValid() into a comparison with a constant.
+    // This maintains a boolean type for the overall expression.
+    auto constant = new IR::Constant(IR::Type::Bits::get(1), 1);
```

Comment:
Why not Cast to bool - it would generate more efficient code in most cases it would be a noop, unlike the comparison.